plugin/manager.go

```diff
@@ -0,0 +1,318 @@
+	}
+
+	pluginFiles, err := ioutil.ReadDir(directory)
+	if err != nil {
+		return fmt.Errorf("error while reading directory %s", err)
+	}
+	for _, f := range pluginFiles {
+		pluginPath := filepath.Join(directory, "./", f.Name())
+		pRaw, err := plugin.Open(pluginPath)
+		if err != nil {
+			return fmt.Errorf("error while opening plugin %s; %s", pluginPath, err)
+		}
+		compatPlugin, err := compat.Wrap(pRaw)
+		if err != nil {
+			return err
+		}
+		if err := m.LoadPlugin(compatPlugin); err != nil {
+			return nil
+		}
+	}
+	return nil
+}
+
+// LoadPlugin loads a compat plugin, exported to sideload plugins for testing purposes
+func (m *Manager) LoadPlugin(compatPlugin compat.Plugin) error {
+	modulePath := compatPlugin.PluginInfo().ModulePath
+	if _, ok := m.plugins[modulePath]; ok {
+		return fmt.Errorf("plugin with module path %s is present at least twice", modulePath)
+	}
+	m.plugins[modulePath] = compatPlugin
+	return nil
+}
+
+// InitializeForUserID initializes all plugin instances for a given user
+func (m *Manager) InitializeForUserID(userID uint) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	user := m.db.GetUserByID(userID)
+	if user != nil {
+		return m.initializeForUser(*user)
+	}
+	return fmt.Errorf("user with id %d not found", userID)
+}
+
+func (m *Manager) initializeForUser(user model.User) error {
+
+	userCtx := compat.UserContext{
+		ID:    user.ID,
+		Name:  user.Name,
+		Admin: user.Admin,
+	}
+
+	for _, p := range m.plugins {
+		if err := m.initializeSingleUserPlugin(userCtx, p); err != nil {
+			return err
+		}
+	}
+
+	for _, app := range m.db.GetApplicationsByUser(user.ID) {
+		if conf := m.db.GetPluginConfByApplicationID(app.ID); conf != nil {
+			_, compatExist := m.plugins[conf.ModulePath]
+			app.Internal = compatExist
+		} else {
+			app.Internal = false
+		}
+		m.db.UpdateApplication(app)
+	}
+
+	return nil
+}
+
+func (m *Manager) initializeSingleUserPlugin(userCtx compat.UserContext, p compat.Plugin) error {
+	info := p.PluginInfo()
+	instance := p.NewPluginInstance(userCtx)
+	userID := userCtx.ID
+
+	pluginConf := m.db.GetPluginConfByUserAndPath(userID, info.ModulePath)
+
+	if pluginConf == nil {
+		var err error
+		pluginConf, err = m.createPluginConf(instance, info, userID)
+		if err != nil {
+			return err
+		}
+	}
+
+	m.instances[pluginConf.ID] = instance
+
+	if compat.HasSupport(instance, compat.Messenger) {
+		instance.SetMessageHandler(redirectToChannel{
+			ApplicationID: pluginConf.ApplicationID,
+			UserID:        pluginConf.UserID,
+			Messages:      m.messages})
+	}
+	if compat.HasSupport(instance, compat.Storager) {
+		instance.SetStorageHandler(dbStorageHandler{pluginConf.ID, m.db})
+	}
+	if compat.HasSupport(instance, compat.Configurer) {
+		if len(pluginConf.Config) == 0 {
+			// The Configurer is newly implemented
+			// Use the default config
+			pluginConf.Config, _ = yaml.Marshal(instance.DefaultConfig())
+			m.db.UpdatePluginConf(pluginConf)
+		}
+		c := instance.EmptyConfig()
```

Comment:
Why not use `DefaultConfig()` to get a pointer to the new Configuration? This not only removes a poilerplate method(`func (c *Plugin) EmptyConfig() interface{} { return new(Config) }`) but also enables plugin to provide default values to some unset keys, eg:
```golang
type Config struct {
    // default: true
    VerifyJWTSignature bool
}

func (c *Plugin) DefaultConfig() interface{} {
     return &Config{true}
}
```
In this case, if this key is missing in the real configuration, it will be set to true by default.

Reply:
Good Idea!